editor/gui/scene_tree_editor.cpp

```diff
@@ -1012,32 +1017,74 @@ void SceneTreeEditor::_rename_node(Node *p_node, const String &p_name) {
 	}
 }
 
-void SceneTreeEditor::_renamed() {
+void SceneTreeEditor::_edited() {
 	TreeItem *which = tree->get_edited();
-
 	ERR_FAIL_COND(!which);
-	NodePath np = which->get_metadata(0);
-	Node *n = get_node(np);
-	ERR_FAIL_COND(!n);
 
-	String new_name = which->get_text(0);
+	if (is_scene_tree_dock && tree->get_next_selected(which)) {
+		List<Node *> to_rename;
+		for (TreeItem *item = which; item; item = tree->get_next_selected(item)) {
+			Node *n = get_node(item->get_metadata(0));
+			ERR_FAIL_NULL(n);
+			to_rename.push_back(n);
+		}
+		ERR_FAIL_COND(to_rename.is_empty());
+
+		EditorUndoRedoManager *undo_redo = EditorUndoRedoManager::get_singleton();
+		undo_redo->create_action(TTR("Rename Nodes"), UndoRedo::MERGE_DISABLE, to_rename.front()->get(), true);
+
+		TreeItem *item = which;
+		for (Node *n : to_rename) {
+			_renamed(item, which, n);
+			item = tree->get_next_selected(item);
+		}
+		which->remove_meta("use_class");
+		undo_redo->commit_action();
+	} else {
+		_renamed(which, nullptr);
+	}
+}
+
+void SceneTreeEditor::_renamed(TreeItem *p_item, TreeItem *p_batch_item, Node *p_node) {
+	Node *n;
+	if (p_node) {
+		n = p_node; // During batch rename the paths may change, so using metadata is unreliable.
+	} else {
+		n = get_node(p_item->get_metadata(0));
+	}
+	ERR_FAIL_NULL(n);
+
+	String new_name;
+	if (p_batch_item) {
+		if (!p_batch_item->get_meta(SNAME("use_class"), false)) {
+			new_name = p_batch_item->get_text(0);
+		}
+	} else {
+		new_name = p_item->get_text(0);
+	}
```

Comment:
once the lower part is handled by `rename_node` (#78292).I think the rest of this function can be integrated into the `_edited` function this will make clearer and more readable code.